[ Melvin Chan ] iP

MelvinCZJ opened the pull request:
# **Duke O.o**
Your No.1 choice of Task Manager!

> “The key is not to prioritize what’s on your schedule, but to schedule your priorities.” – Stephen Covey ([source](https://dansilvestre.com/productivity-quotes/))

Download it from [here!](https://github.com/MelvinCZJ/ip/releases/download/Duke_v0.1/2103T_IP.jar)

**Duke O.o** is a
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,
1. Run it.
2. Add your tasks.
3. Let it manage it for you :grin:

And it is **FREE**!

Features:
[x] Managing tasks
[x] Managing deadlines
[ ] Reminders (coming soon)

---

If you are a Java programmer, you can use it to practice Java too. Here's Duke O.o's main class:
```Java
package Duke.Main;

import Duke.Commands.Command;
import Duke.Commands.CommandEmpty;
import Duke.DukeException.DukeException;
import Duke.System.Parser;
import Duke.System.Storage;
import Duke.System.TaskList;
import Duke.System.Ui;

public class Duke {
    private final Storage storage;
    private TaskList tasks;
    private final Ui ui;

    public Duke(String filePath) {
        ui = new Ui();
        storage = new Storage(filePath);
        try {
            tasks = new TaskList(storage.load());
        } catch (DukeException e) {
            ui.showLoadingError();
            tasks = new TaskList();
        }
    }

    public void run() {
        ui.showGreeting();
        ui.showLine();
        ui.showTutorial();
        ui.showLine();
        boolean isExitCommand = false;
        do {
            Command c = new CommandEmpty();
            try {
                String fullCommand = ui.promptCommand();
                c = Parser.parse(fullCommand);
                c.execute(tasks, ui, storage);
            } catch (DukeException e) {
                ui.showException(e);
            } finally {
                ui.showLine();
            }
            isExitCommand = c.isExit();
        } while (!isExitCommand);
    }

    public static void main(String[] args) {
        new Duke("data").run();
    }
}
```

Commit:
Add Gradle support

MelvinCZJ committed:
Added Greet, Echo, Exit to Duke

MelvinCZJ committed:
Restructured Duke level-1

MelvinCZJ committed:
Added Add, List to Duke

MelvinCZJ committed:
Restructured level-2

MelvinCZJ committed:
Added Task.java

MelvinCZJ committed:
Added ToDo.java, Event.java, Deadline.java

MelvinCZJ committed:
Handles EmptyDescriptionErrors, InvalidinputError

MelvinCZJ committed:
Added Delete

MelvinCZJ committed:
Fix TextUiTesting

MelvinCZJ committed:
Fix coding style related issues

MelvinCZJ committed:
Restructure Duke.java
Add CommandList.java, DateChecker.java

Break down one larger input command into multiple smaller inputs from the user.

The previous versions of Duke accepted one single input to be parsed into multiple different commands.
This led to certain difficulties in parsing those commands.

Breaking down one singular Input command into multiple smaller input mitigates risks of incorrect parsing
of commands.

At each stage of requesting the user for input, the format of the input is explicitly stated to minimize cases of
invalid input fomats. On top of that, each input from the user should also be checked for validity.

Higher levels of interaction between Duke and user can allow users to feel more involved when using Duke,
in order to improve overall user experience.

MelvinCZJ committed:
Add EmptyTask.java, TaskMananger.java

Null checks should not exist! Replaced null checks with EmptyTask check.

In order to make more use of OOP, task related methods have been delegated to the TaskManager class.

MelvinCZJ committed:
Refactor DateChecker.java to DateManager.java

Date manager allows for greater flexibility for printing of dates in different formats.

MelvinCZJ committed:
Add TaskCreator.java

TaskCreator handles creation of task when loading up saved data and also when tasks are created during normal task addition.

MelvinCZJ committed:
Fix minor issues

Issues included: Unused imports, Wrong task constructors used

MelvinCZJ committed:
Add TimeManager.java

TimeManager.java allows Duke to properly process time from a 24-hr format into both the 24-hr and 12-hr formats.
Minor changes were made to allow duke to better process and store date and time.

MelvinCZJ committed:
Merge branch 'branch-level-8'

MelvinCZJ committed:
Bugfix after the merging of level 7 and level 8

Minor incompatibility issues arouse after merging level 7 and level 8. This version of Duke tackles those issues.

MelvinCZJ committed:
Refactor to better make use of OOP

Further separation of codes into their different components allow for smoother troubleshooting process

MelvinCZJ committed:
Package classes into their respective packages.

MelvinCZJ committed:
Add new feature for users to search list by keyword

Added CommandFind.java
Updated Commands enum, Parser, TaskList, Ui to accomodate new feature

MelvinCZJ committed:
Add new feature  to clear list.

Clear allows user to instantly clear the tasklist.
Fixed bugs relating to date and time

MelvinCZJ committed:
Add JavaDoc

Code was tweaked to comply with given coding standard also.

medajet commented on src/main/java/Duke/Main/Duke.java:
```diff
@@ -0,0 +1,51 @@
+        ui.showGreeting();
+        ui.showLine();
+        ui.showTutorial();
+        ui.showLine();
```
Very good naming for this part, since I can tell the code likely prints something to the user without looking at the actual method. Same goes for the other show methods

medajet commented on src/main/java/Duke/System/Parser.java:
```diff
@@ -0,0 +1,121 @@
+
+        String command = params[0].toUpperCase();
+
+        boolean containsCommand = false;
```
Good choice of naming for boolean variable

medajet commented on src/main/java/Duke/System/Ui.java:
```diff
@@ -0,0 +1,109 @@
+                + SPACE + "To view this manual again, input: help\n"
+                + SPACE + "The first word of each command is not case-sensitive!\n"
+                + SPACE + "The format of the date should be in the format DDMMYYYY!\n"
+                + SPACE + "The format of the time should be in the format HHMM!";
```
Looks very neat! You also made sure that single lines don't go too long 👍

medajet commented on src/main/java/Duke/Tasks/Task.java:
```diff
@@ -0,0 +1,47 @@
+package Duke.Tasks;
+
+public abstract class Task {
```
Not related to naming quality guidelines, but I like that you declared this as abstract since it follows OOP principles more closely

medajet commented on src/main/java/Duke/Time/ManagerTime.java:
```diff
@@ -0,0 +1,50 @@
+
+public class ManagerTime {
+    private final String format24;
+    private final String format12;
```
Would naming these two variables as 24hFormat/12hFormat be a better choice for readability? It removes the ambiguity of "format" since it can be both a verb and noun, as just looking at the variable name "format24" for example makes me think that it actually does something to format the time to the 24-h format.

medajet reviewed (commented):
Well done on the naming! Most of the variable names are very clear and have no ambiguity. 👍

zawedcvg commented on src/main/java/Duke/Commands/CommandHelp.java:
```diff
@@ -0,0 +1,13 @@
+import Duke.System.TaskList;
+import Duke.System.Ui;
+
+public class CommandHelp extends Command {
```
Did not use the constructor in this case? I guess you could have used a function instead of making this a class

zawedcvg commented on src/main/java/Duke/Commands/Command.java:
```diff
@@ -0,0 +1,14 @@
+import Duke.System.Ui;
+
+public abstract class Command {
+    public abstract void execute(TaskList tasks, Ui ui, Storage storage) throws DukeException;
```
You could have a constructor here with the inputs being tasks, ui and storage and have the other Commands use this constructor to sort of bring a uniformity in your code.

zawedcvg commented on src/main/java/Duke/Commands/CommandMark.java:
```diff
@@ -0,0 +1,25 @@
+public class CommandMark extends Command{
+    private final int taskNum;
+
+    public CommandMark(String str) throws DukeException {
```
As mentioned in the Command.java, you could call Command's constructor and use that here. This will allow you to run the execute function without any attributes. Again this is more of a design choice I would say,

zawedcvg commented on src/main/java/Duke/System/Parser.java:
```diff
@@ -0,0 +1,121 @@
+        return select(commandNum, params);
+    }
+
+    public static Command select(int num, String[] params) throws DukeException {
```
You could make the function name a bit more informational . Like making it selectTask instead.

zawedcvg commented on src/main/java/Duke/System/TaskList.java:
```diff
@@ -0,0 +1,116 @@
+        return tasks.size();
+    }
+
+    public String tasksAsString() throws DukeException {
```
You could refactor this function to the Ui file instead.  Otherwise really liked your implementation of the TaskList file.

zawedcvg reviewed (commented):
Overall well written code following code style guidelines in almost all the areas. Also really liked the addition of a command tutorial.

MelvinCZJ committed:
Add Gradle support

Duke can now be run using Gradle

MelvinCZJ committed:
Add Checkstyle

Checkstyle is now supported on Gradle

MelvinCZJ committed:
Add suppress for GroovyAssignabilityCheck (Intellij issue)

Prepping for javafx here. (No javafx codes added yet)

MelvinCZJ committed:
Add GUI to Duke

The addition of a GUI to Duke greatly improve user experience!

MelvinCZJ committed:
Refactor command and task classes

The naming conventions of the various command classes and task classes seem uneccessarily confusing. Let's
change the names of these ckasses. E.g. CommandDeadline -> DeadlineCommand

It seems that the '-' infront of the JavaDocs method parameter comments have been
repeated. Let's remove those uneccessary regexes and clean up all JavaDocs as well.

MelvinCZJ committed:
Fix dialog box

The dialog box is not showing all the text in it. Let's fix it.

MelvinCZJ committed:
Merge pull request #2 from MelvinCZJ/A-CodeQuality

Fix dialog box

MelvinCZJ committed:
Merge branch 'A-CodeQuality'

MelvinCZJ committed:
Merge remote-tracking branch 'origin/master'

MelvinCZJ committed:
Add assertions

MelvinCZJ committed:
Remove tutorial on startup

Duke shows the feature list at start up and it fills up the whole screen. The feature list belongs on the user manual so let's remove that.

MelvinCZJ committed:
Merge pull request #3 from MelvinCZJ/branch-A-Assertions

Remove tutorial on startup

MelvinCZJ committed:
Merge remote-tracking branch 'origin/master'

MelvinCZJ committed:
Update GUI

The previous GUI displays all messages in the same color. Let's make the GUI reflect invalid inputs from the user by making those messages a different color.

Images on Duke and user have also been changed in this version.

MelvinCZJ committed:
Add Ui.png

Fixed issue where unmarking of tasks could not be completed as well.

MelvinCZJ committed:
Add user guide

MelvinCZJ committed:
Update README.md

MelvinCZJ committed:
Add JUnit tests
